src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,207 @@
+import classNames from 'classnames/bind';
+import {
+  FC, useContext, useEffect, useRef, useState,
+} from 'react';
+import { Todo } from '../../types/Todo';
+import { AppTodoContext } from '../../contexts/AppTodoContext';
+import {
+  deleteTodo,
+  editTodoTitle,
+  getTodos,
+  toggleTodoStatus,
+} from '../../api/todos';
+import { ErrorType } from '../Error/Error.types';
+import { USER_ID } from '../../react-app-env';
+
+interface Props {
+  todo: Todo,
+  isTempTodo?: boolean,
+}
+
+export const TodoItem: FC<Props> = (
+  { todo, isTempTodo },
+) => {
+  const {
+    todos,
+    setErrorMessage,
+    addProcessingTodo,
+    removeProcessingTodo,
+    setProcessingTodoIds,
+    isTodoProcessing,
+
+    setTodos,
+    setVisibleTodos,
+  } = useContext(AppTodoContext);
+
+  const { title, completed, id } = todo;
+
+  const [editValue, setEditValue] = useState(title);
+  const [isEditAvailable, setIsEditAvailable] = useState(false);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const handleRemoveButton = async () => {
+    addProcessingTodo(id);
+
+    try {
+      await deleteTodo(todo.id);
+      setTodos(await getTodos(USER_ID));
+    } catch {
+      setErrorMessage(ErrorType.DeleteTodoError);
+    } finally {
+      removeProcessingTodo(id);
+    }
+
+    setVisibleTodos(prevVisTodos => (
+      prevVisTodos.filter(prevTodo => prevTodo.id !== todo.id)
+    ));
+  };
+
+  const handleToggleStatus = async () => {
+    addProcessingTodo(id);
+
+    try {
+      const updatedTodo = await toggleTodoStatus(todo);
+      const prevTodoIndex = todos.findIndex(
+        prevTodo => prevTodo.id === todo.id,
+      );
+
+      setVisibleTodos(prevTodos => {
+        const updatedTodos = [...prevTodos];
+
+        updatedTodos[prevTodoIndex] = updatedTodo;
+
+        return updatedTodos;
+      });
+
+      setTodos(await getTodos(USER_ID));
+    } catch {
+      setErrorMessage(ErrorType.UpdateTodoError);
+    } finally {
+      removeProcessingTodo(id);
+    }
+  };
+
+  const handleEditSubmit = async () => {
+    addProcessingTodo(id);
+
+    if (editValue === title) {
+      setProcessingTodoIds({});
+      setIsEditAvailable(false);
+
+      return;
+    }
+
+    if (editValue.trim() === '') {
+      handleRemoveButton();
+      setIsEditAvailable(false);
+
+      return;
+    }
+
+    try {
+      const updatedTodo = await editTodoTitle(id, editValue);
+      const oldTodoIndex = todos.findIndex(prevTodo => prevTodo.id === id);
+
+      setTodos(prevTodos => {
+        prevTodos.splice(oldTodoIndex, 1, updatedTodo);
```

Comment:
```suggestion
      setTodos(prevTodos => {
        const oldTodoIndex = prevTodos.findIndex(prevTodo => prevTodo.id === id);

        prevTodos.splice(oldTodoIndex, 1, updatedTodo);
```

Reply:
Also it can be done in 1 iteration, for example using map (if id is right - replace with new, if no - return same object)